homeassistant/components/qwikswitch.py

```diff
@@ -76,127 +79,98 @@ def is_on(self):
 
-    @callback
-    def async_stop(event):  # pylint: disable=unused-argument
-        """Stop the listener queue and clean up."""
-        nonlocal qsusb
-        qsusb.stop()
-        qsusb = None
-        hass.data[DOMAIN] = {}
-        _LOGGER.info("Waiting for long poll to QSUSB to time out")
+    # Discover all devices in QSUSB
+    if not await qsusb.update_from_devices():
+        raise PlatformNotReady
 
-    hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, async_stop)
+    hass.data[DOMAIN] = qsusb
 
-    # Discover all devices in QSUSB
-    yield from qsusb.update_from_devices()
-    hass.data[DOMAIN]['switch'] = []
-    hass.data[DOMAIN]['light'] = []
+    _new = {'switch': [], 'light': [], 'sensor': sensors}
     for _id, item in qsusb.devices:
-        if (item[QS_TYPE] == QSType.relay and
-                item[QSDATA][QS_NAME].lower().endswith(' switch')):
-            item[QSDATA][QS_NAME] = item[QSDATA][QS_NAME][:-7]  # Remove switch
-            new_dev = QSSwitch(_id, qsusb)
-            hass.data[DOMAIN]['switch'].append(new_dev)
+        if _id in switches:
+            if item[QS_TYPE] != QSType.relay:
+                _LOGGER.warning(
+                    "You specified a switch that is not a relay %s", _id)
+                continue
+            _new['switch'].append(_id)
         elif item[QS_TYPE] in [QSType.relay, QSType.dimmer]:
-            new_dev = QSLight(_id, qsusb)
-            hass.data[DOMAIN]['light'].append(new_dev)
+            _new['light'].append(_id)
         else:
             _LOGGER.warning("Ignored unknown QSUSB device: %s", item)
             continue
-        hass.data[DOMAIN][_id] = new_dev
 
     # Load platforms
-    for comp_name in ('switch', 'light'):
-        if hass.data[DOMAIN][comp_name]:
-            load_platform(hass, comp_name, 'qwikswitch', {}, config)
+    for comp_name in _new:
+        if _new[comp_name]:
+            load_platform(
+                hass, comp_name, DOMAIN, {DOMAIN: _new[comp_name]}, config)
 
     def callback_qs_listen(item):
         """Typically a button press or update signal."""
-        if qsusb is None:  # Shutting down
-            return
-
         # If button pressed, fire a hass event
-        if item.get(QS_CMD, '') in cmd_buttons and QS_ID in item:
-            hass.bus.async_fire('qwikswitch.button.{}'.format(item[QS_ID]))
-            return
+        if QS_ID in item:
+            if item.get(QS_CMD, '') in cmd_buttons:
+                hass.bus.async_fire(
+                    'qwikswitch.button.{}'.format(item[QS_ID]), item)
+                return
+
+            if item[QS_ID] not in qsusb.devices:
+                # Not a standard device in, component can handle packet
+                # i.e. sensors
+                hass.helpers.dispatcher.async_dispatcher_send(
+                    item[QS_ID], item)
 
         # Update all ha_objects
         hass.async_add_job(qsusb.update_from_devices)
```

Comment:
What does this do? Since we're in a callback that is called by `qsusb`, shouldn't `qsusb` already know to update its devices?

Reply:
The QSUSB API has two main methods, listen and devices. These really have no relationship betwen them and I simply use listen to see when there is activity and call update_from_devices.

If I have access to the event loop in listen I could potentially add it there, but then the API becomes less felxible (although only HA uses it at the moment)

Reply:
The aiohttp session is already passed, so passing the event loop too should work from my point of view. But I don't know this library, so you should decide what fits best.

Reply:
Lets keep it like this for now, it clearly shows what it does with the API and the API is a very close to match to the manufacturer's API